spec/rspec/matchers/built_in/include_spec.rb

```diff
@@ -709,7 +709,7 @@ class PseudoHash < SimpleDelegator
 
           domain = Struct.new(:domain) do
             def matches?(url)
-              URI(url).host == self.domain
```

Comment:
If you're going to remove this change the shadowing above.

Reply:
Actually honestly I prefer the self, can we disable the cop instead?

Reply:
I revert to my original then, remove the shadowing!

Reply:
This is an interesting example. I thought of renaming one of the variables, but then my OCD chimed in, and I decided to check if that object *is* treated as a matcher when `failure_message` is defined.

A balancing example:
```
      it 'does treats an object that implements #matches? and failure_message as a matcher' do
        matcher = Struct.new(:value) do
          def matches?(_)
            fail "`matches?` should never be called"
          end

          def failure_message
            'oops'
          end
        end

        expect([matcher.new("rspec.info")]).to include(matcher.new("rspec.info"))

        expect {
          expect([matcher.new("rspec.info")]).to include(matcher.new("foo.com"))
        }.to fail_matching("expected [#{matcher.new("rspec.info").inspect}] to include")
      end
    end
```
and it was still green. This led me to a thought that its `matches?` was never called.
The trail led me to [`values_match?`](https://github.com/rspec/rspec-expectations/blob/a85a42c/lib/rspec/matchers/composable.rb#L67), then to `Matchers.is_a_matcher?`, which returns `true`, and then to [`values_match?`](https://github.com/rspec/rspec-support/blob/cb038de/lib/rspec/support/fuzzy_matcher.rb#L18) that does the matching using `===`, that in turn is only [aliased to `matches?`](https://github.com/rspec/rspec-expectations/blob/a85a42c/lib/rspec/matchers/composable.rb#L45) when `Composable` is included to the matcher.

I played around with [`all_of` argument matcher](https://github.com/toptal/rspec-any_of) recently, tried to implement it as a class rather than to use `RSpec::Matchers.define` DSL, and failed miserably. Just like here, `matches?` never gets called, but `===` does. It felt that there are some differences between how generic and composable matchers can be defined, and I didn't completely understand it. Neither I do now, but at least I've found that bit that `Composable` should be included in argument matchers.

From my point of view, this example is misleading. Am I missing something?
Do you mind if I remove it altogether?

Reply:
Its a regression check of sorts, so it needs to remain, I don't mind if you want to refactor it to be less confusing. This is testing composability of the include matcher and that the object is not treated as a further argument matcher in a nested format, not the matcher definition of the object itself.

Reply:
Done 👍